src/Database/Esqueleto/ForeignKey.hs

```diff
@@ -0,0 +1,245 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TupleSections #-}
+
+module Database.Esqueleto.ForeignKey
```

Comment:
IMO: This module seems really useful, but also might be better suited as a library in addition to Esqueleto. That would allow you to iterate much quicker on the design of it without having to worry about breaking existing Esqueleto features.